packages/react-native-reanimated/src/createAnimatedComponent/NativeEventsManager.ts

```diff
@@ -77,23 +77,38 @@ export class NativeEventsManager implements INativeEventsManager {
     });
   }
 
-  private getEventViewTag() {
+  private getEventViewTag(isComponentUpdated: boolean = false) {
     // Get the tag for registering events - since the event emitting view can be nested inside the main component
     const componentAnimatedRef = this.#managedComponent
-      ._componentRef as AnimatedComponentRef;
-    let newTag: number;
+      ._componentRef as AnimatedComponentRef & { __nativeTag?: number };
     if (componentAnimatedRef.getScrollableNode) {
+      /*
+        In most cases, getScrollableNode() returns a view tag, and findNodeHandle is not required. 
+        However, to cover more exotic list cases, we will continue to use findNodeHandle 
+        for consistency. For numerical values, findNodeHandle should return the value immediately, 
+        as documented here: https://github.com/facebook/react/blob/91061073d57783c061889ac6720ef1ab7f0c2149/packages/react-native-renderer/src/ReactNativePublicCompat.js#L113
+      */
       const scrollableNode = componentAnimatedRef.getScrollableNode();
-      newTag = findNodeHandle(scrollableNode) ?? -1;
-    } else {
-      newTag =
-        findNodeHandle(
-          this.#componentOptions?.setNativeProps
-            ? this.#managedComponent
-            : componentAnimatedRef
-        ) ?? -1;
+      return findNodeHandle(scrollableNode) ?? -1;
     }
-    return newTag;
+    if (this.#componentOptions?.setNativeProps) {
+      // This case ensures backward compatibility with components that
+      // have their own setNativeProps method passed as an option.
+      return findNodeHandle(this.#managedComponent) ?? -1;
+    }
+    if (!isComponentUpdated) {
+      // On the first render of a component, we may already receive a resolved view tag.
+      return this.#managedComponent.getComponentViewTag();
+    }
+    if (componentAnimatedRef.__nativeTag) {
+      // Fast path for native refs
+      return componentAnimatedRef?.__nativeTag;
```

Comment:
For handling events, things get a bit more complicated:
- If the component is a type of list or ScrollView, we want to avoid the potential issue of getting the native tag of a wrapper instead of the ScrollView itself.
- When someone specifies custom `componentOptions`, we should use them because it indicates that they intentionally want to do something different from the norm.
- If it's the first render, whichever method we use, the outcome will be the same.